src/passes/Precompute.cpp

```diff
@@ -714,12 +719,19 @@ struct Precompute
 
 private:
   // Precompute an expression, returning a flow, which may be a constant
-  // (that we can replace the expression with if replaceExpression is set).
-  Flow precomputeExpression(Expression* curr, bool replaceExpression = true) {
+  // (that we can replace the expression with if replaceExpression is set). When
+  // |usedHeapValues| is provided, we use those values instead of the normal
+  // |heapValues| (that is, we do not use the normal heap value cache).
+  Flow precomputeExpression(Expression* curr,
+                            bool replaceExpression = true,
+                            HeapValues* usedHeapValues = nullptr) {
     Flow flow;
     try {
-      flow = PrecomputingExpressionRunner(
-               getModule(), getValues, heapValues, replaceExpression)
+      flow = PrecomputingExpressionRunner(getModule(),
+                                          getValues,
+                                          usedHeapValues ? *usedHeapValues
+                                                         : heapValues,
```

Comment:
Maybe we simplify things by moving this logic to the very top of the function:

```
if (!usedHeapValues) {
  usedHeapValues = &heapValues;
}
```

Reply:
Done.